File: android/core/domain/src/main/java/com/enterprise/core/domain/usecases/LoginUseCase.kt
package com.enterprise.core.domain.usecases

import com.enterprise.core.domain.entities.DomainResult
import com.enterprise.core.domain.entities.User
import com.enterprise.core.domain.interfaces.AnalyticsTracker
import com.enterprise.core.domain.repositories.AuthRepository
import javax.inject.Inject

/** Executes login with credentials, biometric, or passkey and tracks analytics. */
class LoginUseCase @Inject constructor(
    private val authRepository: AuthRepository,
    private val analyticsTracker: AnalyticsTracker
) {
    suspend fun withCredentials(email: String, password: String): DomainResult<User> {
        analyticsTracker.trackEvent("login_attempt", mapOf("method" to "credentials"))
        val result = authRepository.login(email, password)
        when (result) {
            is DomainResult.Success -> analyticsTracker.trackEvent("login_success", mapOf("method" to "credentials"))
            is DomainResult.Failure -> analyticsTracker.trackEvent("login_failure", mapOf("method" to "credentials"))
        }
        return result
    }

    suspend fun withBiometric(): DomainResult<User> {
        analyticsTracker.trackEvent("login_attempt", mapOf("method" to "biometric"))
        return authRepository.loginWithBiometric()
    }

    suspend fun withPasskey(challenge: ByteArray): DomainResult<User> {
        analyticsTracker.trackEvent("login_attempt", mapOf("method" to "passkey"))
        return authRepository.loginWithPasskey(challenge)
    }
}

File: ios/EnterpriseMobileApp/Sources/Core/Domain/UseCases/LoginUseCase.swift
import Foundation

/// Use case for user login with multiple authentication methods.
public struct LoginUseCase: Sendable {
    private let authRepository: AuthRepository
    private let analyticsTracker: AnalyticsTracker

    public init(authRepository: AuthRepository, analyticsTracker: AnalyticsTracker) {
        self.authRepository = authRepository
        self.analyticsTracker = analyticsTracker
    }

    public func execute(email: String, password: String) async -> DomainResult<User> {
        let result = await authRepository.login(email: email, password: password)
        switch result {
        case .success:
            analyticsTracker.trackEvent("login_success", properties: ["method": "credentials"])
        case .failure(let error):
            analyticsTracker.trackEvent("login_failure", properties: ["method": "credentials", "error": "\(error)"])
        }
        return result
    }

    public func executeWithBiometric() async -> DomainResult<User> {
        let result = await authRepository.loginWithBiometric()
        switch result {
        case .success:
            analyticsTracker.trackEvent("login_success", properties: ["method": "biometric"])
        case .failure(let error):
            analyticsTracker.trackEvent("login_failure", properties: ["method": "biometric", "error": "\(error)"])
        }
        return result
    }

    public func executeWithPasskey(challenge: Data) async -> DomainResult<User> {
        let result = await authRepository.loginWithPasskey(challenge: challenge)
        switch result {
        case .success:
            analyticsTracker.trackEvent("login_success", properties: ["method": "passkey"])
        case .failure(let error):
            analyticsTracker.trackEvent("login_failure", properties: ["method": "passkey", "error": "\(error)"])
        }
        return result
    }
}
